File: resources/android/src/com/victorycodedev/plugins/toastkit/manager/ToastKitManager.kt
package com.victorycodedev.plugins.toastkit.manager

import android.os.Handler
import android.os.Looper
import androidx.compose.runtime.mutableStateListOf
import com.victorycodedev.plugins.toastkit.bridge.ToastKitBridgeNormalizer
import com.victorycodedev.plugins.toastkit.bridge.ToastKitEventDispatcher
import com.victorycodedev.plugins.toastkit.model.*

internal data class ToastKitShowResult(val id: String, val accepted: Boolean)

internal object ToastKitManager {
    val visible = mutableStateListOf<ToastKitConfiguration>()
    val rendered = mutableStateListOf<ToastKitConfiguration>()
    val exiting = mutableStateListOf<String>()
    private val states = linkedMapOf<String, ToastKitState>()
    private val waiting = ArrayDeque<String>()
    private val timers = mutableMapOf<String, Runnable>()
    private val main = Handler(Looper.getMainLooper())
    private val identityLock = Any()
    private val activeIds = mutableSetOf<String>()
    private val uniqueToId = mutableMapOf<String, String>()

    fun show(configuration: ToastKitConfiguration): ToastKitShowResult {
        val result = synchronized(identityLock) {
            configuration.uniqueKey?.let { key ->
                uniqueToId[key]?.let { return@synchronized ToastKitShowResult(it, false) }
            }
            if (!activeIds.add(configuration.id)) return@synchronized ToastKitShowResult(configuration.id, false)
            configuration.uniqueKey?.let { uniqueToId[it] = configuration.id }
            ToastKitShowResult(configuration.id, true)
        }
        if (!result.accepted) return result
        onMain {
            states[configuration.id] = ToastKitState(configuration)
            if (canAdmit(configuration)) admit(configuration.id) else waiting.addLast(configuration.id)
        }
        return result
    }

    fun update(id: String, changes: Map<String, Any>) = onMain {
        val state = states[id] ?: return@onMain
        if (state.terminated) return@onMain
        val updated = try { ToastKitBridgeNormalizer.applyChanges(state.configuration, changes) } catch (_: IllegalArgumentException) { return@onMain }
        states[id] = state.copy(configuration = updated)
        if (state.visible) {
            replaceVisible(updated)
            if (changes.containsKey("duration") || changes.containsKey("persistent")) schedule(updated)
        }
    }

    fun updateUnique(key: String, changes: Map<String, Any>): String {
        val id = resolveUnique(key)
        update(id, changes)
        return id
    }

    fun dismiss(id: String, reason: String = "programmatic") = onMain { terminate(id, reason) }

    fun dismissUnique(key: String): String {
        val id = resolveUnique(key)
        dismiss(id)
        return id
    }

    fun dismissAll() = onMain {
        states.keys.toList().forEach { terminate(it, "programmatic", promote = false) }
        waiting.clear()
        promote()
    }

    fun action(id: String, actionId: String) = onMain {
        val state = states[id] ?: return@onMain
        if (state.terminated) return@onMain
        ToastKitEventDispatcher.action(id, actionId)
        terminate(id, "action")
    }

    private fun canAdmit(configuration: ToastKitConfiguration): Boolean {
        if (exiting.isNotEmpty()) return false
        val current = visible.mapNotNull { states[it.id]?.configuration }
        return when (configuration.strategy) {
            ToastKitStrategy.QUEUE -> current.isEmpty()
            ToastKitStrategy.STACK -> current.none { it.strategy == ToastKitStrategy.QUEUE } &&
                current.count { it.strategy == ToastKitStrategy.STACK } < configuration.maxVisible
        }
    }

    private fun admit(id: String) {
        val state = states[id] ?: return
        if (state.terminated || state.visible) return
        states[id] = state.copy(visible = true)
        visible.add(state.configuration)
        rendered.add(state.configuration)
        schedule(state.configuration)
        ToastKitEventDispatcher.shown(id)
    }

    private fun schedule(configuration: ToastKitConfiguration) {
        timers.remove(configuration.id)?.let(main::removeCallbacks)
        val duration = configuration.durationMs ?: return
        val task = Runnable { terminate(configuration.id, "timeout") }
        timers[configuration.id] = task
        main.postDelayed(task, duration)
    }

    private fun terminate(id: String, reason: String, promote: Boolean = true) {
        val state = states[id] ?: return
        if (state.terminated) return
        states[id] = state.copy(terminated = true)
        timers.remove(id)?.let(main::removeCallbacks)
        waiting.remove(id)
        visible.removeAll { it.id == id }
        exiting.add(id)
        val exitDuration = if (!state.visible) 0L else if (state.configuration.nativeAndroid) 160L else when (state.configuration.animation) {
            ToastKitAnimation.SPRING, ToastKitAnimation.BOUNCE -> 450L
            ToastKitAnimation.REVEAL -> 280L
            else -> 260L
        }
        main.postDelayed({
            rendered.removeAll { it.id == id }
            exiting.remove(id)
            releaseIdentity(state.configuration)
            if (promote) promote()
        }, exitDuration)
        states.remove(id)
        ToastKitEventDispatcher.dismissed(id, reason)
    }

    private fun resolveUnique(key: String): String = synchronized(identityLock) {
        uniqueToId[key] ?: throw IllegalArgumentException("Unique toast [$key] is not active.")
    }

    private fun releaseIdentity(configuration: ToastKitConfiguration) = synchronized(identityLock) {
        activeIds.remove(configuration.id)
        configuration.uniqueKey?.let { key ->
            if (uniqueToId[key] == configuration.id) uniqueToId.remove(key)
        }
    }

    private fun promote() {
        var progressed: Boolean
        do {
            progressed = false
            val id = waiting.firstOrNull() ?: break
            val state = states[id]
            if (state == null || state.terminated) {
                waiting.removeFirst()
                progressed = true
            } else if (canAdmit(state.configuration)) {
                waiting.removeFirst()
                admit(id)
                progressed = true
            }
        } while (progressed)
    }

    private fun replaceVisible(configuration: ToastKitConfiguration) {
        val index = visible.indexOfFirst { it.id == configuration.id }
        if (index >= 0) visible[index] = configuration
        val renderedIndex = rendered.indexOfFirst { it.id == configuration.id }
        if (renderedIndex >= 0) rendered[renderedIndex] = configuration
    }

    private inline fun onMain(crossinline operation: () -> Unit) {
        if (Looper.myLooper() == Looper.getMainLooper()) operation() else main.post { operation() }
    }
}

File: resources/ios/Sources/Manager/ToastKitManager.swift
import Foundation
import SwiftUI

@MainActor
final class ToastKitManager: ObservableObject {
    static let shared = ToastKitManager()

    @Published private(set) var visible: [ToastKitConfiguration] = []
    private var interactiveFrames: [String: CGRect] = [:]
    private var states: [String: ToastKitState] = [:]
    private var waiting: [String] = []
    private var timers: [String: Task<Void, Never>] = [:]

    func show(_ configuration: ToastKitConfiguration) {
        guard states[configuration.id] == nil else { return }
        states[configuration.id] = ToastKitState(configuration: configuration)
        if canAdmit(configuration) { admit(configuration.id) } else { waiting.append(configuration.id) }
    }

    func update(_ id: String, changes: [String: Any]) throws {
        guard let state = states[id], !state.terminated else { return }
        let updated = try ToastKitBridgeNormalizer.applying(changes, to: state.configuration)
        var next = state
        next.configuration = updated
        states[id] = next
        if state.visible {
            if let index = visible.firstIndex(where: { $0.id == id }) { visible[index] = updated }
            if changes.keys.contains("duration") || changes.keys.contains("persistent") { schedule(updated) }
        }
    }

    func dismiss(_ id: String, reason: String = "programmatic") { terminate(id, reason: reason) }

    func dismissAll() {
        for id in Array(states.keys) { terminate(id, reason: "programmatic", shouldPromote: false) }
        waiting.removeAll()
        promote()
    }

    func pressAction(toastId: String, actionId: String) {
        guard let state = states[toastId], !state.terminated else { return }
        ToastKitEventDispatcher.action(toastId, actionId: actionId)
        terminate(toastId, reason: "action")
    }

    func updateInteractiveFrames(_ frames: [String: CGRect]) {
        interactiveFrames = frames
    }

    func containsInteractivePoint(_ point: CGPoint) -> Bool {
        interactiveFrames.values.contains { $0.insetBy(dx: -8, dy: -8).contains(point) }
    }

    private func canAdmit(_ configuration: ToastKitConfiguration) -> Bool {
        switch configuration.strategy {
        case .queue: return visible.isEmpty
        case .stack:
            return !visible.contains(where: { $0.strategy == .queue }) &&
                visible.filter { $0.strategy == .stack }.count < configuration.maxVisible
        }
    }

    private func admit(_ id: String) {
        guard var state = states[id], !state.terminated, !state.visible else { return }
        state.visible = true
        states[id] = state
        withAnimation(animation(for: state.configuration)) {
            visible.append(state.configuration)
        }
        schedule(state.configuration)
        ToastKitEventDispatcher.shown(id)
    }

    private func schedule(_ configuration: ToastKitConfiguration) {
        timers.removeValue(forKey: configuration.id)?.cancel()
        guard let milliseconds = configuration.durationMilliseconds else { return }
        timers[configuration.id] = Task { [weak self] in
            try? await Task.sleep(for: .milliseconds(milliseconds))
            guard !Task.isCancelled else { return }
            self?.dismiss(configuration.id, reason: "timeout")
        }
    }

    private func terminate(_ id: String, reason: String, shouldPromote: Bool = true) {
        guard var state = states[id], !state.terminated else { return }
        state.terminated = true
        states[id] = state
        timers.removeValue(forKey: id)?.cancel()
        waiting.removeAll { $0 == id }
        withAnimation(animation(for: state.configuration)) { visible.removeAll { $0.id == id } }
        states.removeValue(forKey: id)
        ToastKitEventDispatcher.dismissed(id, reason: reason)
        releaseIdentity(for: state)
        if shouldPromote { promote() }
    }

    private func releaseIdentity(for state: ToastKitState) {
        guard state.visible else {
            ToastKitUniqueRegistry.shared.release(state.configuration)
            return
        }
        let milliseconds = switch state.configuration.animation {
        case .spring, .bounce: 450
        case .reveal: 280
        default: 260
        }
        Task { @MainActor in
            try? await Task.sleep(for: .milliseconds(milliseconds))
            ToastKitUniqueRegistry.shared.release(state.configuration)
        }
    }

    private func promote() {
        while let id = waiting.first {
            guard let state = states[id], !state.terminated else { waiting.removeFirst(); continue }
            guard canAdmit(state.configuration) else { break }
            waiting.removeFirst()
            admit(id)
        }
    }

    private func animation(for toast: ToastKitConfiguration) -> Animation {
        switch toast.animation {
        case .fade, .slide, .scale, .reveal: return .easeInOut(duration: 0.2)
        case .snap: return .spring(response: 0.22, dampingFraction: 0.62)
        case .spring: return .spring(response: 0.35, dampingFraction: 0.78)
        case .pop: return .spring(response: 0.28, dampingFraction: 0.58)
        case .bounce: return .spring(response: 0.48, dampingFraction: 0.55)
        }
    }
}
